### PR TITLE
Improve prime checks

### PR DESCRIPTION
**Pull request by kroeckx**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by mspncp** [comment]
> Based on research by Jake Massimo and Kenneth Paterson

Is there some reference to a publication?

**Comment by paulidale** [comment]
I've not been able to find a reference to the specific publication either.

**Comment by kroeckx** [comment]
It's explained in the commit message, but it's speed.

(I'm still working on things like comments.)

**Comment by kroeckx** [comment]
On Fri, Jun 28, 2019 at 10:30:46PM -0700, Matthias St. Pierre wrote:
> > Based on research by Jake Massimo and Kenneth Paterson
> 
> Is there some reference to a publication?

Last time I asked about it, they've submitted it to ACM CCS and
didn't plan to submit it anywhere else until they got review back.
So I currently can't give you a reference to it.

**Comment by mspncp** [comment]
> Last time I asked about it, they've submitted it to ACM CCS and didn't plan to submit it anywhere else until they got review back. So I currently can't give you a reference to it.

So is it newer work than the two papers [mentioned her](https://dblp.uni-trier.de/pers/hd/m/Massimo:Jake) and there is no preprint online?

**Comment by kroeckx** [comment]
> So is it newer work than the two papers [mentioned her](https://dblp.uni-trier.de/pers/hd/m/Massimo:Jake) and there is no preprint online?

Yes. I expect them make a preprint available soon.

**Comment by kroeckx** [comment]
I've pushed a new commit, still need to update the documentation.

**Comment by t8m** [comment]
Do we really want to deprecate the BN_is_prime_fasttest_ex()? Although it is fairly convoluted use-case - let's say 3rd party wants to implement their own RSA key gen based on the OpenSSL BIGNUMs, it would be useful for them to have the fasttest where they can set a low number of MR rounds.

File: crypto/bn/bn_prime.c [hunk 1]
```diff
@@ -191,15 +208,23 @@ int BN_generate_prime_ex(BIGNUM *ret, int bits, int safe,
 }
 #endif
 
+#if !OPENSSL_API_1_1_0
```
[the remark this answers is not in the document]
Comment (kroeckx) [reply crypto/bn/bn_prime.c]: Probably, fixed.

**Comment by kroeckx** [comment]
I think the only use case for doing less Miller-Rabin iterations is prime generation, and you can argue that it doesn't hurt to do more iterations, it's just a little slower. All the internal calls that are left are about prime generation.

**Comment by paulidale** [comment]
> Do we really want to deprecate the BN_is_prime_fasttest_ex()

I believe so.  The easier we make prime testing to use, the less mistakes can be made.  Ideally there would be one _is this a prime?_ function.  Sure there are performance reasons for having many, but the others ought to be internal to the libraries.

I'm not sure about the `BN_is_prime_ex2` name but can live with it.  I'll suggest: `OSSL_is_prime` as an alternative and that we move to deprecate the other APIs eventually.

**Comment by t8m** [comment]
More alternatives for the name would be BN_test_prime() or BN_check_prime() :)

**Comment by kroeckx** [comment]
How would you feel that I've also switched to the new function for
prime generation, and that all BN_is_prime* functions actually did
the same thing, and so ignore the number of iterations (or use it
as a minimum.)

**Comment by paulidale** [comment]
I'd be fine with this.

Let's wait for others to comment in case the performance loss is problematic.

**Comment by kroeckx** [comment]
I would at least like to change the minimum security level of the
generated primes to match the 128 bit level, while we now do, 64,
80, 112, 128, 192 or 256 depending on the prime size.

**Comment by richsalz** [comment]
FWIW, I think those changes are great, @kroeckx I think making keygen a bit slower for more security is the right trade-off; keygen isn't common.

**Comment by t8m** [comment]
So in my completely ad-hoc artificial testing on my machine the original keygen for 8k RSA keys took 8.5 seconds in average and with the prime checks with 128 MR rounds it took 10 seconds in average. I think this is acceptable.

**Comment by t8m** [comment]
In general the variance of the times was higher than the slowdown.

**Comment by JBeatz** [comment]
> > Based on research by Jake Massimo and Kenneth Paterson
> 
> Is there some reference to a publication?

Hi all, author Jake Massimo here. Just closing the loop on this by providing links to the publication (now that it is in conference proceedings) for anyone who is interested.

CCS 2020: https://dl.acm.org/doi/10.1145/3372297.3417264 
ePrint: https://eprint.iacr.org/2020/065

Thanks for all your work and cooperation on this, it's been a pleasure.

**Pull request closed without merging**